src/chunked_scatter/safe_scatter.py

```diff
@@ -0,0 +1,244 @@
+# Copyright (c) 2019 Leiden University Medical Center
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import argparse
+import math
+from typing import Generator, Iterable, List
+
+from .chunked_scatter import common_parser, region_lists_to_scatter_files
+from .parsers import BedRegion, file_to_regions
+
+
+def merge_regions(regions: Iterable[BedRegion]
+                  ) -> Generator[BedRegion, None, None]:
+    """
+    Merge regions that overlap or are exactly adjacent
+    :param regions: An iterable of possibly overlapping regions
+    :return: a generator of merged regions
+    """
+    merged_region = None
+    for region in regions:
+        if merged_region is None:
+            merged_region = region
+        else:
+            if (merged_region.contig == region.contig and
+                    merged_region.end >= region.start and
+                    region.end >= merged_region.start):
+                start = min(merged_region.start, region.start)
+                end = max(merged_region.end, region.end)
+                merged_region = BedRegion(merged_region.contig, start, end)
+            else:
+                yield merged_region
+                merged_region = region
+    if merged_region:
+        yield merged_region
+
+
+def sum_regions(regions: List[BedRegion]):
+    """ Calculate the total length of all regions """
+    return sum(len(region) for region in regions)
```

Comment:
This number will not be entirely correct when there are overlapping regions. Is that intentional?
`return sum(len(region) for region in merge_regions(regions))` is also an option.

Reply:
There should be no overlapping regions, since the regions are generated based on the number of requested chunks. So it should not be possible to get overlapping regions into this function.